### PR TITLE
blog: updated E2E best practices

### PR DESCRIPTION
**Pull request by pohly**
"Writing good E2E tests" was already updated a while ago in https://github.com/kubernetes/community/pull/7021. As suggested there (https://github.com/kubernetes/community/pull/7021#issuecomment-1374333094), we should bring this update to the attention of more contributors, hence this blog post.

/cc @aojea @jberkus

### CODE REVIEW

**Review by xmcqueen — APPROVED** [review]
excellent
/lgtm

**Comment by pohly** [comment]
/assign @mrbobbytables 

For approval. Let's get this published before KubeCon, then folks can chat with me about it there.

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:176]:
```suggestion
for that code (see next section). It is better than `ginkgo.AfterEach` because
```

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:181]: maybe you should introduce this at the beginning of the section, that the framework creates a test namespace to avoid test pollution ... I don't know if this behavior of "test namespaces" is well known outside of kubernetes

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:183]: also, if you create objects that are going to take some times to be completed delated, I had this mistakes with "terminating pods", I set a grace period of 300 seconds and the pod blocks the garbage collectors for that time

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:186]: indeed this is a common mistake

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:280]:
```suggestion
When waiting for something to happen and you need to do asynchronous assertions, use a reasonable timeout. Without it, a
```

Comment (aojea) [reply content/en/blog/2023/e2e-testing-best-practices.md:280]: o asynchronous checks, I think people is familiar with this term

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:283]: " On the other hand,"  here is misleading, I think that you may express the too main problems
- short timeout, the test will flake, per example if the CI is slow
- long timeout, the test may hide underline issues, per example, if there are some races with other components and eventually the condition pass

The thing with timeouts, is that you also should define what is the expected time you consider valid for a an operation to succeed, e2e are not only functiona,  i.e. creating a pod and it takes more than 10 minutes to run should not pass because that environment is too busy, or Services can not take more than 1 minute in program the dataplance, ... timeouts are also important to set the upper limits for some behaviors

Comment (aojea) [reply content/en/blog/2023/e2e-testing-best-practices.md:283]: the key is to obtain the right balance, a timeout that doesn't flake and that the time is considerable acceptable for that operation

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:304]: it also accepts contexts :)

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [inline content/en/blog/2023/e2e-testing-best-practices.md:369]: this is important, unfortunately I'm afraid we are not doing it as much as we should

**Comment by aojea** [comment]
LGTM
overall good summary of the work done and the expectations for e2e tests, it will be good to promote this blog for reviewers of the different SIGs , so we can have consistency on all the e2e tests.

I left some comments that are not blockers and if necessary can be follow ups

/assign @sftim 

you need docs people IIRC for approval

**Comment by pohly** [comment]
/hold

I'll follow up on some of the suggestions before this gets merged.

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (pohly) [reply content/en/blog/2023/e2e-testing-best-practices.md:283]: There are two problems with too long timeouts:
- a feature is broken and some expected state will never occur, but the test needs to run till it times out waiting for that state
- a feature is normally supposed to work within a certain time frame and for some reason is taking too long

The problem is that we don't have good enough control over the performance of the clusters that we test against, nor do many features have any solid "must work within XYZ seconds". Solving that problem goes beyond what we can solve right now.

I'll clarify the first point and add a comment about the second.

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (pohly) [reply content/en/blog/2023/e2e-testing-best-practices.md:297]: I'll explain that the amount of information should depend on how the E2E suite was invoked.

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (pohly) [reply content/en/blog/2023/e2e-testing-best-practices.md:304]: Which is part of "all criteria", right? So no need to change anything in the text.

**Comment by pohly** [comment]
/hold cancel

PR updated, ready for another LGTM and approval.

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (sftim) [inline content/en/blog/2023/e2e-testing-best-practices.md:4]: For a publication date, how about 2023-04-12?

**Review by sftim — COMMENTED** [review]
/lgtm cancel

The publication date should be in the future at the time we merge it. However, it looks otherwise OK.

**Comment by sftim** [comment]
BTW, approval is SIG ContribEx blog team.

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (aojea) [reply content/en/blog/2023/e2e-testing-best-practices.md:311]: yeah, I was just a personal comment to show my +1 to this function

**Comment by aojea** [comment]
/lgtm

Thanks

File: content/en/blog/2023/e2e-testing-best-practices.md
Comment (pohly) [reply content/en/blog/2023/e2e-testing-best-practices.md]: Works for me, updated accordingly: https://github.com/kubernetes/contributor-site/compare/846fba3cd23f37924ea22e687ad015d838a4a7fd..7f1a7e9808047ed5c84b577428ccdfab209c4d4e

**Review by mrbobbytables — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**
